enterprise/models.py

```diff
@@ -1906,34 +1979,64 @@ def __repr__(self):
         return self.__str__()
 
 
-class LicensedEnterpriseCourseEnrollment(TimeStampedModel):
+class EnterpriseFulfillmentSource(TimeStampedModel):
     """
-    An Enterprise Course Enrollment that is enrolled via a license.
-
-    .. no_pii:
+    Base class for enterprise subsidy fulfillments
     """
+    class Meta:
+        abstract = True
 
-    license_uuid = models.UUIDField(
-        primary_key=False,
+    LICENSE_FULFILLMENT_TYPE = "license"
+    LEARNER_CREDIT_FULFILLMENT_TYPE = "learner_credit"
+    COUPON_CODE_FULFILLMENT_TYPE = "coupon_code"
+
+    fulfillment_type_choices = [
+        (LICENSE_FULFILLMENT_TYPE, "License"),
+        (LEARNER_CREDIT_FULFILLMENT_TYPE, "Learner Credit"),
+        (COUPON_CODE_FULFILLMENT_TYPE, "Coupon Code"),
+    ]
+
+    uuid = models.UUIDField(
+        primary_key=True,
         editable=False,
-        null=False
+        null=False,
+        default=uuid4,
+    )
+
+    fulfillment_type = models.CharField(
+        max_length=128,
+        choices=fulfillment_type_choices,
+        default=LICENSE_FULFILLMENT_TYPE,
+        help_text=f"Subsidy fulfillment type, can be one of: "
+        f"{(LICENSE_FULFILLMENT_TYPE, LEARNER_CREDIT_FULFILLMENT_TYPE, COUPON_CODE_FULFILLMENT_TYPE)}"
+    )
+
+    enterprise_course_entitlement = models.OneToOneField(
+        EnterpriseCourseEntitlement,
+        blank=True,
+        null=True,
+        related_name="%(app_label)s_%(class)s_entitlement_related",
+        on_delete=models.deletion.CASCADE,
+        help_text=_(
+            "The course entitlement the associated subsidy is for."
+        )
     )
 
     enterprise_course_enrollment = models.OneToOneField(
         EnterpriseCourseEnrollment,
-        blank=False,
```

Comment:
Validation is done now to ensure either an enrollment or entitlement are present